### DISCUSSION TITLE
Cannot use Spanner Client inside a PythonOperator because of incompatibility issues

### DISCUSSION THREAD

**Discussion by piffarettig** [apache/airflow#30497] [Q&A]
### Apache Airflow Provider(s)

google

### Versions of Apache Airflow Providers

`apache-airflow-providers-google==8.11.0`

### Apache Airflow version

2.5.2

### Operating System

Debian GNU/Linux

### Deployment

Docker-Compose

### What happened

I'm using a PythonOperator that just executes some SQL and make the results available to a subsequent task via the usage of xcoms. This is because there's no default operator available for Cloud Spanner that gives me that for free. 

Why do I need to do this? Because I need to DROP all Spanner tables that follow a given prefix, so it's not that I can anticipate the table names. They're dynamic (I only know about their prefix).

However, my code fails when I do `with database.snapshot() as snapshot:`.  I get:

> TypeError: Expected int, bytes, or behavior, got <class 'grpc_gcp_pb2.ApiConfig'>

That's the exact same error mentioned [here](https://github.com/apache/beam/issues/22454).

And given my Airflow version is 2.5.2, the Google Provider requires:

- `google-api-core>=2.7.0,<3.0.0`
- `google-cloud-spanner>=1.10.0,<2.0.0`

Which resolves to:

- `google-api-core==2.8.2`
- `google-cloud-spanner==1.19.3`

And those are the exact same dependency versions mentioned in the issue above. These dependencies don't seem to change in more recent Airflow versions.

### How to reproduce

Generate the following task in a DAG:

```python
def my_task() -> PythonOperator:
    task_id = "generate-spanner-drop-statements"
    return PythonOperator(
        task_id=task_id,
        python_callable=_delete_old_spanner_tables,
        provide_context=True,
        op_kwargs={
            "table_prefix": "your table prefix",
            "project_id": "your project id",
            "instance_id": "your instance id",
            "database_id": "your database id"
        },
    )
    
def _delete_old_spanner_tables(table_prefix: str, project_id: str, instance_id: str, database_id: str, **kwargs):
    hook = SpannerHook()
    database = hook.get_database(
        project_id=project_id,
        instance_id=instance_id,
        database_id=database_id
    )
    
    if not database:
        raise AirflowException(
            f"The Cloud Spanner database '{database_id}' in project '{project_id}' "
            f"and instance '{instance_id}' is missing."
        )

    # Select statement that gives me the drop statements of the tables and indexes I want to delete
    with open("sql/drop_old_spanner_tables.sql", "r") as f:
        contents = f.read()
    statement = contents.replace("{{ table_prefix }}", table_prefix)

    with database.snapshot() as snapshot:
        result = snapshot.execute_sql(statement)
        rows = list()
        for row in result:
            rows.append(row)
        task_instance = kwargs['task_instance']
        task_instance.xcom_push(key='drop_statements', value=rows)
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by eladkal** [comment]
If your question is why we don't use newer versions for google libraries it's because newer versions are conflicting. There is a campaign to update them : https://github.com/apache/airflow/pull/30067

If you need newer versions for a specific task you should use `PythonVirtualenvOperator`.

**Comment by piffarettig** [comment]
I'm not sure I understand your answer @eladkal. These are the already default versions that come with Airflow `2.5.2`, it's not that I'm trying to use newer versions. In other words, the already provided versions are conflicting.

Also, how do you specifically suggest using `PythonVirtualenvOperator` to overcome this?

**Comment by piffarettig** [comment]
I ask the latter question because it doesn't seem possible to use `PythonVirtualenvOperator` passing `provide_context=True` and not getting pickle serialization errors. Therefore, I just cannot access the task instance to do `xcom_push()`. See [here](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html#id1)

**Comment by eladkal** [comment]
If pypi allows to install together then they are not conflicting.
If there are bugs with these libs then ask mantainers to resolve them. As explained in Airflow constraints docs we provide a stable installation (in terms of installation 
is successful) we do not update constraints after generated. If you need diffrent versions then create your own constraints file and use it. That is very OK.

**Comment by potiuk** [comment]
I am afraid that until https://github.com/apache/airflow/pull/30067 is solved, there is nothing you can do short of simply doing what Elad mentioned. The problem is that google provider's libraries limit the core-api/spanner to versions that have this problem (apparently). And the only solution to use it is to wait until it gets resolved (which will be some time in May most likely). Simply the libraries used there are far behind - this is being fixed now.

The things you can do:

* use Bash Opeator and google cloud CLI to run some of your operations with spanner
* implement the PythonVirtualenv mentioned by Elad. You **should** be able to get virtualenv with "apache-airflow==2.5.2"  (no extras) any future version of spanner library you choose and custom code (you can start with copying the code from the Spanner's Hook/Operator to run whatever you need there. If you use install the exact same "apache-airflow" version and the same Python version - it should work without serialization issues as long as you pass all arguments as plain serializable things, and you avold to pass any non-serializable objects back/forth. The "serialization" errors you mention likely come from those. Also you do not need to do xcom_push in your PythonVirtualenvOperator. You can return any (serializable) value and it should be automatically pushed to xcom for the operator (after serialize/deserialize) is done successfully.
* you can implement your own docker image with set of libraries you find useful and run it with DockerOpratora
* if you go to Kubernetes, you cans use KubernetesPodOperator similarly

The options are summarized here: https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#handling-conflicting-complex-python-dependencies

It is what it is, this is the unfortunate situation google provider is now in, and short of solving the dependencies in it, not much can be done about it in provider itself - you need to workaround it in one of the ways described above.
